Cdodge/restricted oauth2 applications

chrisndodge opened the pull request:
@efagin here is the PR from Microsoft (under its Enterprise Contributor Agreement) to implement what we discussed as an initial step towards authorization-code based OAuth2 Provider support.

Here are the key implementation points:

- A creation of an appropriately styled/branded Authorization page which overrides the unstyled version in django-oauth-toolkit

- A new data model (table) called RestrictedApplication which merely indicates which django-oauth-toolkit Application should be marked as 'Restricted'. A standard Django Admin page is exposed for SysAdmins to manage such RestrictedApplications

- Any authorization-code generated for a RestrictedApplication should be auto expired (here we set the expiration to Jan. 1, 1970)

- Any JWT token generated for a RestrictedApplication should be auto expired (likewise set to expire Jan. 1, 1970)

- A new OAuth2 scope called "identity" which limits the JWT claims to be the minimal set (namely preferred_username)

- Corresponding unit-tests to assert the above

Since this PR represents a small step of a OAuth2 SSO handshake, I've created a small test client web application - written in Flask - which can be found at https://github.com/chrisndodge/test-edx-platform-oauth2.  This will allow you to see the flow as well as the new Authorization web page. When your reviewers are ready, I'll forward - via email - additional instructions and OAuth secrets you'll need to run it.

FYI, @jbarciauskas, @chenriksson

edx-webhook commented:
Thanks for the pull request, @chrisndodge! I've created [OSPR-1484](https://openedx.atlassian.net/browse/OSPR-1484) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

shaunagm commented:
jenkins run all

efagin commented:
Thanks Chris! Tagging @nasthagiri, @jcdyer, and @clintonb to review this PR, as they're the subject-matter experts here and are all aware of the project. (If not, you all can feel free to reach out to me directly for more context.)

One question for you, @chrisndodge: have you read OEP-4 yet, which specifically details how we'd like to approach application authorization in the platform? http://open-edx-proposals.readthedocs.io/en/latest/oep-0004.html. You may find that helpful in understanding where we'd like to go long term with scope governance, such as naming conventions (e.g. `identity:read` instead of just `identity` to clarify that this is just a read-only scope).

As an aside, I want to make sure we're not overlapping or otherwise conflating OpenID Connect scopes/terms - and also using them where appropriate so we're not reinventing any wheels. Just something to keep an eye on, as it looks like `identity` is fairly similar to the `profile` and `email` OIDC default scopes at a glance.

efagin commented:
(and: apologies for the delay, was offline since Friday afternoon)

cdodge-am commented:
jenkins run all

clintonb commented on lms/templates/oauth2_provider/authorize.html:
```diff
@@ -0,0 +1,54 @@
+
+                        <div class="control-group">
+                            <div class="controls">
+                                <button type="submit" class="btn btn-authorization-cancel" name="cancel"/>Cancel</button><button type="submit" class="btn btn-authorization-allow" name="allow" value="Authorize"/>Allow</button>
```
Mark the button text for translation.

clintonb commented on lms/templates/oauth2_provider/authorize.html:
```diff
@@ -0,0 +1,54 @@
+                    </form>
+
+                {% else %}
+                    <h2>Error: {{ error.error }}</h2>
```
Mark the `Error:` label for translation.

clintonb commented on lms/templates/oauth2_provider/authorize.html:
```diff
@@ -0,0 +1,54 @@
+{% load theme_pipeline %}{% load sekizai_tags i18n configuration %}{% load url from future %}{% load staticfiles %}
+
+{% block title %}
+  {% block pagetitle %}{% endblock %} | {% trans "Authorize" %} | {% platform_name %}
```
The `pagetitle` block is not used, and can probably be removed.

clintonb replied:
Yep, this should definitely be removed.

clintonb commented on lms/templates/oauth2_provider/authorize.html:
```diff
@@ -0,0 +1,54 @@
+{% extends "main_django.html" %}
+{% with online_help_token="oauth2_provider_authorization" %}
+{% load theme_pipeline %}{% load sekizai_tags i18n configuration %}{% load url from future %}{% load staticfiles %}
```
Put each of these `load` calls on a separate line to make this more legible.

clintonb replied:
Actually...none of these seem to be used besides `i18n`. The unused libraries should be removed.

clintonb commented on lms/templates/oauth2_provider/authorize.html:
```diff
@@ -0,0 +1,54 @@
+{% extends "main_django.html" %}
+{% with online_help_token="oauth2_provider_authorization" %}
```
This doesn't seem to be used.

clintonb commented on lms/envs/common.py:
```diff
@@ -446,8 +446,16 @@
+        # the user's preferred_username. This can be used by
+        # 3rd party apps to link accounts. No other data is exchanged if
+        # this is the only scope requested
+        'identity': 'Receive basic identity information, such as your public username',
```
The profile scope will provide the username and real name. See https://github.com/edx/edx-oauth2-provider/blob/master/edx_oauth2_provider/oidc/handlers.py#L140-L170.

chrisndodge commented:
Thanks @clintonb, I should be able to address those quickly.

The intent around the 'identity' scope was to have a scope which exposes the bare minimum that we need. Also, I wasn't sure if there are a lot of 'profile' scoped access_tokens/JWT claims out in the wild and if we might want a distinct scope to isolate new claims from older ones.

We're happy to switch to just using 'profile', but I will want to update that description string that is displayed to end users.

clintonb commented:
@chrisndodge I believe most of our tokens include the profile scope since most are generated for OpenID Connect.

clintonb commented:
I created a test client using https://github.com/ging/oauth2-example-client, and checked the UI and access token. My only concern is that the token includes all scopes. https://github.com/evonove/django-oauth-toolkit/pull/395 is supposed to address application-specific scopes.

Results
```json
{
  "scopes": [
    "read",
    "write",
    "profile",
    "email",
    "identity"
  ],
  "administrator": true,
  "aud": "lms-key",
  "iss": "https://oauth-test.sandbox.edx.org/oauth2",
  "preferred_username": "edxapp",
  "name": "edxapp",
  "exp": 0,
  "iat": 1475697879,
  "email": "edxapp@example.com",
  "sub": "e5de420d61dbb35b3097bb25b7a9b59a"
}
```

![screen shot 2016-10-05 at 3 33 07 pm](https://cloud.githubusercontent.com/assets/910510/19129665/18b2f222-8b16-11e6-9819-660d50d7339b.png)

chrisndodge commented:
@clintonb regarding all scopes, as per the spec, your test client should request 'scope=xxxxx' in the query string parameters. Yes, the PR into the main django-oauth-toolkit (evonove/django-oauth-toolkit#395) regarding limiting which applications can request which scopes has not been merged. I can certainly add some code to make 'restricted oauth applications' only be granted a subset of scopes, but that's duplicative to that pending PR.

Also, @efagin, can we come to a conclusion regarding the 'identity' scope. I can either a) rename that to be 'identity:read' (or something similar) or b) drop it and we use the existing 'profile' scope. My intent to introduce that new scope was a) not expose any PII in our JWT claims, since we don't need it  and b) I wasn't sure of the proliferation of existing 'profile' access_tokens/JWT claims and if we want to segregate out these new claims or not.

Either is fine with me and is a small adjustment.

clintonb commented:
@chrisndodge I made a new request and confirmed the scope limitation is working. 👍

chrisndodge commented on lms/djangoapps/oauth_dispatch/dot_overrides.py:
```diff
@@ -3,9 +3,37 @@
+
+    if is_application_restricted:
+        # put the expire timestamp into the beginning of the epoch
+        instance.expires = datetime(1970, 1, 1, tzinfo=utc)
```
@clintonb I tried this but I'm getting access_tokens rows in the database with the expiry year of '0001-01-01' which seems a bit abnormal and I'm concerned certain clients might not properly parse this.

Any issue on keeping this to be 1970-01-01 which is - as I understand - the beginning of the Unix epoch?

clintonb replied:
No issues. This can stay as-is.

chrisndodge commented:
@clintonb I addressed your code bits of feedback - although I'm not 100% sure on the import ordering rules. Note sure why github is not marking your comments as 'out of date' to indicate which bits have been addressed.

I think remaining question is regarding this 'identity' scope. Perhaps edX could give some feedback on preferred route and I'll conform. Then I can update sandbox (although it appears you are running your own) and squash commits.

chrisndodge commented:
@clintonb unfortunately, I see oauth_dispatch has recently been moved in the last few hours (by @nasthagiri ?) from lms/djangoapps to openedx/core/djangoapps so all these changes are in conflict with upstream master :-(

clintonb commented:
I'll let @efagin weigh in on scopes. Note that `profile` returns username and real name.

chrisndodge commented:
@chenriksson given the recent conflicting relocation of the oauth_dispatch folder, we'll need rebase our fork against the tip of upstream master so I can resolve merge conflicts.

clintonb commented:
@chrisndodge you may be the victim of the hackathon taking place today and tomorrow. @nasthagiri what are your plans for the OAuth-related apps? When will things be settled so that @chrisndodge can rebase and merge his work?

nasthagiri commented:
@chrisndodge Sorry about the conflicts.  We are working to refactor and relocate code in edx-platform to their more appropriate locations.  There's one more oauth-related PR that's about to land (hopefully it doesn't affect you): https://github.com/edx/edx-platform/pull/13659

chrisndodge commented:
Thanks @nasthagiri, I guess it'll be best to wait for your hackathon to end and we'll do a rebase. Late afternoon tomorrow, perhaps?

chenriksson commented:
@chrisndodge sounds good. Ping me tomorrow pm if you want me to sync our fork with upstream.

chrisndodge commented:
@clintonb any idea why my last few commits haven't triggered a Jenkins build?

clintonb commented:
@chrisndodge I have no idea; but, I suspect it may be due to the conflicts. I recall a few changes to Jenkins to avoid running unnecessary builds. @benpatterson can confirm/deny/explain.

benpatterson commented:
jenkins ok to test

benpatterson commented:
@chrisndodge i'm not sure why you went unrecognized. perhaps you've been pulled out of the edx org in the past few days?

efagin commented:
My preference would be to use the existing `profile` scope instead of creating a new `identity` scope, and update the user-facing phrasing for existing scopes to be much more clear than the existing defaults. @clintonb what do you think?

@chrisndodge in the interest of time and containment, I'd recommend we tackle the phrasing in a separate PR and just keep this one to the "restricted" client application code. I'd like the phrasing to come from edX's legal team directly so we can take that as an action item. I understand it's a blocker for you guys. (If you have a preferred phrasing for `profile`, we can take that into account.)

nasthagiri commented:
@chrisndodge Yes, the hackathon is until later this afternoon.  After which, we may still have a few lingering PRs pending to be merged.  However, I don't anticipate further PRs (from our project) that will affect auth beyond today.

Once again, sorry for the timing of this and the merge conflicts.  If it will help, I'd be glad to rebase your PR with master for you later this evening or on Monday.  Feel free to let me know how you'd like to proceed.

clintonb commented:
@chrisndodge @efagin my preference is to use the existing scope.

chrisndodge commented:
@nasthagiri thanks for your offer to help, but I think I need to resolve the conflicts because we are working from our own fork and you don't have push rights. I did a dry rebase run yesterday and it didn't look too bad. I'll let you know if I run into anything.

chrisndodge commented:
@clintonb @efagin thanks, I'll conform to using the existing 'profile' scope. Should be a small change. I'll defer to others on my team regarding any desired messaging to end-users explaining the 'profile' scope and if there are any required timelines to getting final messaging.

chrisndodge commented:
@clintonb @nasthagiri @efagin OK, I rebased, resolved conflicts (and moved new files into openedx/core/djangoapps/oauth_dispatch), and squashed commits.

I had a green build this afternoon, so I'm hoping this one will pass too.

I've updated our sandbox for internal testing with our MPP OAuth2 client application. We'll probably do a round of manual testing Tuesday.

HOWEVER, @efagin, I think the "profile scopes" text on the Authorization page isn't really that good. Do you have an ETA when you think edX Legal can provide an alternative?

chrisndodge commented:
@nasthagiri hrm, something's not right after the rebase and I had a clean build before the rebase. It seems like the migrations that I added into oauth_dispatch are not running on the python unit tests on Jenkins. I cannot repro this locally - running migrations locally in my environment adds the new table.

I see in common.py we aren't including oauth_dispatch in INSTALLED_APPS but rather 'openedx.core.djangoapps.oauth_dispatch.apps.OAuthDispatchAppConfig' which I don't fully understand and I'd rather not start changing things willy-nilly. But it was similar before your refactoring.

Any ideas?

It might be helpful if you pull this branch and try running migrations locally. Does it create a new oauth_dispatch_restrictedapplication table?

chrisndodge commented:
jenkins run python

chrisndodge commented:
jenkins test python

chrisndodge commented:
@nasthagiri OK, maybe failed migrations was a transient error.

Am I able to use the "jenkins run xxxxx" comments to trigger rebuilds? Or is that an edX only internal capability?

chrisndodge commented:
jenkins run all

chrisndodge commented:
@clintonb @nasthagiri I think I believe I know why unit tests are failing post-refactoring. oauth_dispatch is now in a shared library (openedx/core/djangoapps) and thus are running in both LMS and CMS unit test executions. However in CMS settings we're not setting oauth_dispatch in the list of INSTALLED_APPS. I see @nasthagiri in https://github.com/edx/edx-platform/pull/13658 used a pattern of making conditional unittest execution on certain test cases. I've just applied the same pattern. Should see in an hour or so if that is the right approach.

efagin commented:
@chrisndodge please take a look at the existing template from the old provider: https://github.com/edx/edx-platform/blob/master/lms/templates/provider/authorize.html. It contains the phrasing we've previously set up for those scopes. I'd say, let's start there and try to minimize the differences between authorization screens as a goal of this initial change.

I've kicked off a separate discussion to try to figure out what we want the terminology to be moving forward, but "Read your user profile" is at least better than "Profile scope."

chrisndodge commented:
@clintonb @nasthagiri I'm no longer getting build failures on the oauth_dispatch tests, but I've twice gotten "[Errno 12] Cannot allocate memory" failures on some github related tests. Please advise....

@efagin I've conformed to that other profile scope message

nasthagiri commented:
@chrisndodge Unfortunately, the out-of-memory issues are sporadically biting all of us in edx-platform.  The test-eng team is trying to diagnose the issue - most likely a memory leak.

chrisndodge commented:
Thanks @nasthagiri. Unfortunately, it doesn't seem like I can issue 'jenkins test xxx' commands, so I have to commit a trivial change, rebase to squash commits, and push. Is it correct that Open edX contributors cannot trigger builds of specific test groups via 'jenkins test xxxx'?

gsong commented:
@chrisndodge I've asked our test engineers to add you to the Jenkins whitelist. I'll let you know when that happens.

nasthagiri commented on lms/envs/common.py:
```diff
@@ -450,9 +450,14 @@
 }
+# This is required for the migrations in oauth_dispatch.models
+# otherwise it fails saying this attribute is not present in Settings
+OAUTH2_PROVIDER_APPLICATION_MODEL = 'oauth2_provider.Application'
```
Since this has been added here, can remove it from [lms/envs/test.py](https://github.com/edx/edx-platform/blob/master/lms/envs/test.py#L587) and [cms/envs/test.py](https://github.com/edx/edx-platform/blob/master/cms/envs/test.py#L337).

chrisndodge replied:
Sorry, I'm not sure I understand the request here. Can you explain a bit more?

nasthagiri replied:
Sorry for not being clear.  The lines linked in the `test.py` file should be removed with this PR - since the addition of this line in `common.py` makes the `test.py` code unnecessary.

chrisndodge replied:
Ah, now I understand. Got it, thanks!

nasthagiri commented on openedx/core/djangoapps/oauth_dispatch/tests/test_dot_adapter.py:
```diff
@@ -38,6 +43,16 @@ def setUp(self):
+            client_id='restricted-client-id',
+        )
+        restricted_app = RestrictedApplication.objects.create(application=self.restricted_client)
+        self.assertEqual(unicode(restricted_app), u"<RestrictedApplication '{name}'>".format(
```
Typically we test assertions within a test method outside of the `setUp` method.  Can you move this to a (tiny) test method?

gsong commented:
@chrisndodge You've been whitelisted for Jenkins tests. See <https://openedx.atlassian.net/wiki/display/TE/Jenkins+Guide> for more details.

chrisndodge commented on openedx/core/djangoapps/oauth_dispatch/dot_overrides.py:
```diff
@@ -61,6 +88,16 @@ def save_bearer_token(self, token, request, *args, **kwargs):
+            # must be automatically expired. So let's update
+            # the token dictionary, so set the expires_in field be
+            # on Jan. 1, 1970
+            token['expires_in'] = -int(time.time())
```
This is because the underlying DOT library does not re-read the AccessToken after the presave, so the DOT library returns an 'expires_in' which does not reflect the override. We decided not to fork the underlying DOT library, so we need to include this seemingly redundant code

chrisndodge commented:
Thank you @gsong

chrisndodge commented:
@nasthagiri thanks. I believe I've addressed your pieces feedback that I understood.

chrisndodge commented:
jenkins test python

chrisndodge commented:
jenkins run python

chrisndodge commented:
jenkins run python

chrisndodge commented:
jenkins run python

nasthagiri commented on openedx/core/djangoapps/oauth_dispatch/dot_overrides.py:
```diff
@@ -61,6 +88,16 @@ def save_bearer_token(self, token, request, *args, **kwargs):
 
+        is_application_restricted = RestrictedApplication.objects.filter(application=request.client).exists()
+
+        if is_application_restricted:
```
Need to update the docstring of this method.  The current docstring refers only to overrides for the client-credentials grant.

nasthagiri commented on openedx/core/djangoapps/oauth_dispatch/dot_overrides.py:
```diff
@@ -61,6 +88,16 @@ def save_bearer_token(self, token, request, *args, **kwargs):
+            # must be automatically expired. So let's update
+            # the token dictionary, so set the expires_in field be
+            # on Jan. 1, 1970
+            token['expires_in'] = -time_since_epoch()
```
Rather than assuming the `expires_in` value in the database equals the epoch time, can we read the value from the database here and set it according to the database value?

nasthagiri commented on openedx/core/djangoapps/oauth_dispatch/tests/test_views.py:
```diff
@@ -63,12 +111,13 @@ def _post_body(self, user, client, token_type=None):
     """
     def setUp(self):
         self.url = reverse('access_token')
+        self.login_with_access_token_url = reverse("login_with_access_token")
```
This line shouldn't be needed here since it's set in the `setUp` method of `AccessTokenLoginMixin`.

chrisndodge replied:
For some reason - maybe it's some pythonic thing I'm not aware of - I can't seem to get a def setup() in the mixin to execute. If I simply make it an attribute on the class, then it works fine.

nasthagiri replied:
@chrisndodge You should declare the Mixin class before the concrete class in the list of base classes.  And the `setUp` implementation on the Mixin class should make sure to call `super` (even if it doesn't have any base classes).  This will ensure that all class ancestors are called through the `super`-chain.

chrisndodge replied:
Thanks, will do!

nasthagiri commented on openedx/core/lib/token_utils.py:
```diff
@@ -98,3 +98,12 @@ def encode(self, payload):
         return jwt.encode(payload, secret, algorithm=algorithm)
+
+
+def time_since_epoch():
```
Thanks for adding this utility function - however, there's only caller to this function.
I was actually thinking instead about adding a common place where we define the expiration time for DOT-issued `RestrictedApplication` tokens.  This way, if we ever choose to change that value, it would be in one place.  And also, it isn't hard-coded in multiple places in the code.

nasthagiri commented on openedx/core/djangoapps/oauth_dispatch/tests/test_views.py:
```diff
@@ -109,6 +174,47 @@ def test_jwt_access_token(self, client_attr):
+            should_be_expired=True
+        )
+
+    def test_restricted_acces_token(self):
```
typo: access

nasthagiri commented:
Nit: Can you update the commit message as well?
1. privileges is misspelled (or it's using a deprecated spelling)
2. rewrite to be within the recommended 72 character limit.

For example: `add restricted DOT OAUTH2 client with reduced privileges`

chrisndodge commented:
jenkins run python

chrisndodge commented:
jenkins run bokchoy

chrisndodge commented:
@nasthagiri I think I got through your last batch of comments.

I'm still having problems with flakey builds. Not sure if there's something I could be doing better.

nasthagiri commented:
(For future reference, please don't squash your commits while the PR review is still in progress. It makes the review process longer - especially for a PR of moderate size.)

nasthagiri commented on openedx/core/djangoapps/oauth_dispatch/dot_overrides.py:
```diff
@@ -61,6 +83,17 @@ def save_bearer_token(self, token, request, *args, **kwargs):
+            # Since RestrictedApplications will override the DOT defined expiry, so that access_tokens
+            # are always expired, we need to re-read the token from the database and then calculate the
+            # expires_in (in seconds) from what we stored in the database. This value should be a negative
+            #value, meaning that it is already expired
```
Nit: To validate the assumption in your comment, you can add an assert to make sure it's negative.

chrisndodge replied:
Good idea, will do

chrisndodge replied:
done

nasthagiri commented on openedx/core/djangoapps/oauth_dispatch/dot_overrides.py:
```diff
@@ -61,6 +83,17 @@ def save_bearer_token(self, token, request, *args, **kwargs):
+
+            access_token = AccessToken.objects.get(token=token['access_token'])
+            utc_now = datetime.utcnow().replace(tzinfo=utc)
+            token['expires_in'] = (access_token.expires - utc_now).total_seconds()
```
+1 by reading from the model, we can be assured that the `expires_in` value is set to the true value.

nasthagiri commented on openedx/core/djangoapps/oauth_dispatch/migrations/0001_initial.py:
```diff
@@ -0,0 +1,22 @@
+    ]
+
+    operations = [
+        migrations.CreateModel(
```
FYI @edx/devops on this upcoming migration.  It shouldn't be a problem - since it's a new table.  However, keep in mind that this table has a foreign relationship to the OAuth2 `Application` table.

chrisndodge commented:
@nasthagiri sure thing re: squashing commits. I was asked up above to squash commits, but I'll hold off for now.

nasthagiri commented on openedx/core/djangoapps/oauth_dispatch/models.py:
```diff
@@ -0,0 +1,37 @@
+        )
+
+    @classmethod
+    def set_access_token_as_expired(cls, access_token):
```
+1 on encapsulating this within the `RestrictedApplication` class.

Nit: Can you do the same for all code that verifies that a restricted token is expired?  In other words, one should only find one occurrence of "1970" in this PR.

nasthagiri commented on openedx/core/djangoapps/oauth_dispatch/tests/test_views.py:
```diff
@@ -3,23 +3,60 @@
+    Shared helper class to assert proper access levels when using access_tokens
+    """
+
+    login_with_access_token_url = reverse("login_with_access_token")
```
Defining this as a class variable won't work in CMS tests since "login_with_access_token" is not defined when `ENABLE_OAUTH2_PROVIDER` is unset.  You can set it in the `setUp` method (you'll need to correct the Mixin class order - add the Mixin before the TestCase class).

This is what is causing the CMS test failures:
https://build.testeng.edx.org/job/edx-platform-test-subset/617147/console

nasthagiri commented:
looks great.  Just waiting for jenkins now before giving thumbs.

nasthagiri commented:
Feel free to squash whenever.  :)

chrisndodge commented:
jenkins run python

nasthagiri commented:
argh! with these python out-of-memory failures..

benpatterson commented:
@nasthagiri we have been working on the OOM issues. It is related to the spurious coverage failures and has not been an easy one to solve. But it's actively being worked on.

FWIW

[I was watching this thread because I was mentioned earlier]

chrisndodge commented:
@nasthagiri looks like we now have a passing build

nasthagiri commented:
@benpatterson thanks for the update.  it definitely sounds like a hard problem.  best of luck.

nasthagiri commented:
@chrisndodge Coverage looks good as well.
You're good to merge from my perspective.
@clintonb - any other comments from you?  The PR has changed since you last reviewed it.

chrisndodge commented:
Thanks for the help/support @nasthagiri!

efagin commented:
Thanks all for the reviews and effort here. @clintonb could you please give the latest changes a once-over and merge when ready?

jzoldak commented:
Leaving a note here that this PR was reverted in #13745 because the merge to master broke database setup of the cms python unit tests.

chrisndodge commented:
Thanks for the heads up. I'll rebase again from master and see if we can reproduce on the PR branch. I had rebased last Friday, but perhaps something has changed on master since then.

jzoldak commented:
@chrisndodge if it's related to something merged recently to master, you probably want to first look at #13732 which we merged yesterday. We are now by default not building the test database via migrations, but instead through models.
